In [1]:
# confusion matrix for Retrieval Algorithm 1
TP_1 = 25
FP_1 = 15
FN_1 = 5
TN_1 = 55


# confusion matrix for Retrieval Algorithm 2
TP_2 = 20 
FP_2 = 10
FN_2 = 10
TN_2 = 60


In [2]:
precision_1 = TP_1 / (TP_1 + FP_1)

precision_2 = TP_2 / (TP_2 + FP_2)

In [3]:
recall_1 = TP_1 / (TP_1 + FN_1)

recall_2 = TP_2 / (TP_2 + FN_2)

In [4]:
f1_score_1 = 2 * (precision_1 * recall_1) / (precision_1 + recall_1)

f1_score_2 = 2 * (precision_2 * recall_2) / (precision_2 + recall_2)

In [5]:
balanced_accuracy_1 = 0.5 * ((TP_1 / (TP_1 + FN_1)) + (TN_1 / (TN_1 + FP_1)))

balanced_accuracy_2 = 0.5 * ((TP_2 / (TP_2 + FN_2)) + (TN_2 / (TN_2 + FP_2)))

In [6]:
print(f"Algorithm 1 - Precision: {precision_1:.4f}, Recall: {recall_1:.4f}, F1-Score: {f1_score_1:.4f}, Balanced Accuracy: {balanced_accuracy_1:.4f}")
print(f"Algorithm 2 - Precision: {precision_2:.4f}, Recall: {recall_2:.4f}, F1-Score: {f1_score_2:.4f}, Balanced Accuracy: {balanced_accuracy_2:.4f}")

Algorithm 1 - Precision: 0.6250, Recall: 0.8333, F1-Score: 0.7143, Balanced Accuracy: 0.8095
Algorithm 2 - Precision: 0.6667, Recall: 0.6667, F1-Score: 0.6667, Balanced Accuracy: 0.7619


#### Question 1

Retrieval Algorithm 1
| Actual Outcome | Predicted Positive | Predicted Negative |
|---------------|------------------|------------------|
| **Positive**  | TP = 25           | FN = 5           |
| **Negative**  | FP = 15           | TN = 55          |

Retrieval Algorithm 2
| Actual Outcome | Predicted Positive | Predicted Negative |
|---------------|------------------|------------------|
| **Positive**  | TP = 20           | FN = 10          |
| **Negative**  | FP = 10           | TN = 60          |


#### Question 2

Balanced Accuracy and F1- score are calculated as follows:

$\text{Balanced Accuracy} = \frac{1}{2} (\text{Sensitivity} + \text{Specificity})$

$F1 = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}}$


**For Retrieval Algorithm 1:**

- True Positives (TP) = 25

- False Negatives (FN) = 5

- False Positives (FP) = 15

- True Negatives (TN) = 55

- TNR (Specificity) = $\frac{TN}{TN + FP} = \frac{55}{70} = 0.79$

- Balanced Accuracy = $\frac{1}{2} (\frac{25}{30} + \frac{55}{70}) = 0.81$

- $F1 = 2 \times \frac{0.625 \times 0.8333}{0.625 + 0.8333} = 0.71$


**For Retrieval Algorithm 2:**

- True Positives (TP) = 20

- False Negatives (FN) = 10

- False Positives (FP) = 10

- True Negatives (TN) = 60

- TNR (Specificity) = $\frac{TN}{TN + FP} = \frac{60}{70} = 0.86$

- Balanced Accuracy = $\frac{1}{2} (\frac{20}{30} + \frac{60}{70}) = 0.76$

- $F1 = 2 \times \frac{0.6667 \times 0.6667}{0.6667 + 0.6667} = 0.67$


Since we are only interested in performance on the negative class, the best metric is True Negative Rate (TNR) instead of Balanced Accuracy or F1-Score. Balanced Accuracy considers both positive and negative classes, but the question specifically asks about the negative class. F1-score does not take True Negatives into account at all, so it is a poor choice for evaluation in this case. The only metric that would directly measure how well the model identifies negative examples is TNR (Specificity). Since Retrieval Algorithm 2 has the higher TNR (0.86 vs. 0.79 for Algorithm 1), Retrieval Algorithm 2 is the better model for identifying negative examples.
Thus, neither the instructor nor the friend is correct. Instead, the best metric is True Negative Rate (TNR), which was not originally suggested by either of them. 

#### Question 3

No, neither my friend nor my instructor was correct because Balanced Accuracy and F1-score evaluate overall performance rather than focusing solely on the negative class. Since the question specifically asks about performance on the negative class, TNR (Specificity) is the correct metric to use. If we only look at Balanced Accuracy and F1-Score, they would both suggest that Algorithm 1 is better since both the Balanced Accuracy Score and the F1 Score are higher for Retrieval Algorithm 1. These scores however measure the **overall** performance of the classifier, and not specifically the performance of the classifier on the negative class. In this case, Algorithm 1 may be a better classifier overall, but Algorithm 2 has better performance on the negative class. Since neither my friend's nor my instructor's suggestion leads me to the correct conclusion, the metrics they suggested to use were incorrect in this situation.

#### Question 4

The best metrics to evaluate performance on the negative class are:

- True Negative Rate (TNR) / Specificity: Measures how well negative examples are correctly classified.
- False Positive Rate (FPR): Measures how often negative examples are incorrectly classified as positive.

These metrics directly measure how well the model performs on the negative class, unlike Balanced Accuracy or F1-score.